File: 2025-03-March-GFG-POTD/March 15 - Coin Change (Minimum Coins)/Solution.java
//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.lang.*;
import java.util.*;


// } Driver Code Ends

class Solution {

    public int minCoins(int coins[], int sum) {
        // If sum is 0, no coins are needed
        if (sum == 0) return 0;
        
        int n = coins.length;
        int dp[] = new int[sum + 1];
        
        // Initialize dp array with a large value
        Arrays.fill(dp, sum + 1);
        dp[0] = 0; // Base case: 0 coins needed for sum 0
        
        for (int i = 1; i <= sum; i++) {
            for (int coin : coins) {
                if (coin <= i) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        
        return (dp[sum] == sum + 1) ? -1 : dp[sum];
    }
}


//{ Driver Code Starts.

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            int k = Integer.parseInt(br.readLine());
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;
            Solution obj = new Solution();
            int res = obj.minCoins(arr, k);

            System.out.println(res);

            System.out.println("~");
        }
    }
}

// } Driver Code Ends

File: 2025-03-March-GFG-POTD/March 15 - Coin Change (Minimum Coins)/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minCoins(vector<int>& coins, int sum) {
        if (sum == 0) return 0;

        vector<int> dp(sum + 1, sum + 1);
        dp[0] = 0;

        for (int i = 1; i <= sum; i++) {
            for (int coin : coins) {
                if (coin <= i) {
                    dp[i] = min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        return (dp[sum] == sum + 1) ? -1 : dp[sum];
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        int k;
        cin >> k;
        vector<int> coins(k);
        for (int i = 0; i < k; i++) {
            cin >> coins[i];
        }
        int sum;
        cin >> sum;
        Solution obj;
        cout << obj.minCoins(coins, sum) << endl;
    }
    return 0;
}
